src/orion/storage/track.py

```diff
@@ -0,0 +1,464 @@
+        """See `~orion.core.worker.trial.Trial`"""
+        self.storage.metadata['end_time'] = value
+
+
+class Track(BaseStorageProtocol):
+    """Implement a generic protocol to allow Orion to communicate using
+    different storage backend
+
+    Parameters
+    ----------
+    uri: str
+        Track backend to use for storage; the format is as follow
+         `protocol://[username:password@]host1[:port1][,...hostN[:portN]]][/[database][?options]]`
+
+    """
+
+    def __init__(self, uri):
+        self.uri = uri
+        self.options = parse_uri(uri)['query']
+
+        self.client = TrackClient(uri)
+        self.backend = self.client.protocol
+        self.project = None
+        self.group = None
+        self.current_trial = None
+
+        self.objective = self.options.get('objective')
+        assert self.objective is not None, 'An objective should be defined!'
+
+    def create_experiment(self, config):
+        """Insert a new experiment inside the database"""
+        if self.project is None:
+            self.project = self.backend.get_project(Project(name=config['name']))
+
+            if self.project is None:
+                self.project = self.backend.new_project(Project(name=config['name']))
+
+        self.group = self.backend.new_trial_group(
+            TrialGroup(
+                name=str(uuid.uuid4()),
+                project_id=self.project.uid,
+                metadata=to_json(config)
+            )
+        )
+
+        config['_id'] = self.group.uid
+        return config
+
+    def update_experiment(self, experiment, where=None, **kwargs):
+        """Update a the fields of a given trials
+
+        Parameters
+        ----------
+        experiment: Experiment
+            Experiment object to update
+
+        where: Optional[dict]
+            constraint experiment must respect
+
+        **kwargs: dict
+            a dictionary of fields to update
+
+        Returns
+        -------
+        returns true if the underlying storage was updated
+
+        """
+        pass
+
+    def fetch_experiments(self, query):
+        """Fetch all experiments that match the query"""
+        return self.backend.fetch_groups(query)
+
+    def register_trial(self, trial):
+        """Create a new trial to be executed"""
+        stamp = datetime.datetime.utcnow()
+        trial.status = 'new'
+        trial.submit_time = stamp
+
+        metadata = dict()
+        metadata['params_types'] = {remove_leading_slash(p.name): p.type for p in trial. params}
+        metadata['submit_time'] = to_json(trial.submit_time)
+        metadata['end_time'] = to_json(trial.end_time)
+        metadata['worker'] = trial.worker
+        metadata['metric_types'] = {remove_leading_slash(p.name): p.type for p in trial.results}
+        metadata['metric_types'][self.objective] = 'objective'
+        metadata['heartbeat'] = stamp
+
+        metrics = defaultdict(list)
+        for p in trial.results:
+            metrics[p.name] = p.value
```

Comment:
A result value is a float or an integer, not a list. The defaultdict(list) won't have lists as values.